core/src/avm2/vector.rs

```diff
@@ -0,0 +1,357 @@
+//! Storage for AS3 Vectors
+
+use crate::avm2::activation::Activation;
+use crate::avm2::object::Object;
+use crate::avm2::value::Value;
+use crate::avm2::Error;
+use gc_arena::Collect;
+use std::cmp::{max, min};
+use std::ops::{Index, RangeBounds};
+use std::slice::SliceIndex;
+
+/// The vector storage portion of a vector object.
+///
+/// Vector values are restricted to a single type, decided at the time of the
+/// construction of the vector's storage. The type is determined by the type
+/// argument associated with the class of the vector. Vector holes are
+/// evaluated to a default value based on the type of the vector.
+///
+/// A vector may also be configured to have a fixed size; when this is enabled,
+/// attempts to modify the length fail.
+#[derive(Collect, Debug, Clone)]
+#[collect(no_drop)]
+pub struct VectorStorage<'gc> {
+    /// The storage for vector values.
+    ///
+    /// While this is structured identically to `ArrayStorage`, the role of
+    /// `None` values is significantly different. Instead of being treated as a
+    /// hole to be resolved by reference to the prototype chain, vector holes
+    /// are treated as default values. The actual default changes based on the
+    /// type contained in the array.
+    storage: Vec<Option<Value<'gc>>>,
```

Comment:
I was under the impression that this would cause problems somewhere with double-borrows or being called in a place that doesn't have an `activation`, but apparently not!